File: Ch01/Ex1_09.hs
{-
 - Ex1_09.hs
 - Kevin C. Baird
 - SICP in Haskell
 -
 - Define both a recursive and interative process for addition
 - (Exercise 1.9)
 -
 -}

module Ex1_9 where

inc :: Int -> Int
inc x = x + 1

dec :: Int -> Int
dec x = x - 1

recursiveAdd :: Int -> Int -> Int
recursiveAdd a b
  | a == 0    = b
  | otherwise = inc (recursiveAdd (dec a) b)

iterativeAdd :: Int -> Int -> Int
iterativeAdd a b
  | a == 0    = b
  | otherwise = iterativeAdd (dec a) (inc b)

File: Ch01/ex1_09.erl
-module(ex1_09).
-author("Kevin C. Baird").
-purpose("SICP in Erlang: Exercise 1.9 - Define both recursive & iterative process for addition").
-export([inc/1, dec/1, recursive_add/2, iterative_add/2]).

inc(X) -> X + 1.
dec(X) -> X - 1.

recursive_add(0, B) -> B;
recursive_add(A, B) -> inc(recursive_add( dec(A), B ) ).

iterative_add(0, B) -> B;
iterative_add(A, B) -> iterative_add( dec(A), inc(B) ).

